### PR TITLE
fix: do not call `<shell> -i` for non-interactive `flox activate --`

### LINKED ISSUE #666: Non-interactive `activate` invokes interactive shell
`flox activate -- COMMAND` currently runs interactive rc scripts, which is surprising behavior as the user cannot interact with the shell.
We should drop the `-i` here, but doing so would require changing our initialization scripts:
https://github.com/flox/flox/blob/866b5fde83c455b026f8af78bfd333e6a04fc91d/cli/flox/src/commands/environment.rs#L406

### PR DESCRIPTION
**Pull request by ysndr**
fixes #666

### CODE REVIEW

File: cli/tests/environment-activate.bats [hunk 1]
```diff
@@ -151,10 +151,6 @@ env_is_activated() {
   # SHELL=zsh USER="$REAL_USER" run -0 bash -c "echo exit | $FLOX_CLI activate --dir $PROJECT_DIR";
   SHELL=zsh USER="$REAL_USER" NO_COLOR=1 run -0 expect -d "$TESTS_DIR/activate/hook.exp" "$PROJECT_DIR"
   assert_output --partial "Welcome to your flox environment!"
-
-  SHELL=zsh USER="$REAL_USER" NO_COLOR=1 run $FLOX_BIN activate --dir "$PROJECT_DIR" -- :
-  assert_success
-  assert_output --partial "Welcome to your flox environment!"
```
Comment (ysndr) [inline cli/tests/environment-activate.bats:157]: We don't run init scripts anymore so aliases set in the rc files are not available in the subshell

File: cli/flox/src/commands/environment.rs [hunk 2]
```diff
@@ -565,28 +565,57 @@ impl Activate {
...
+        run_args: Vec<String>,
+        shell: ShellType,
+        exports: HashMap<&str, String>,
+        activation_path: PathBuf,
+    ) -> anyhow::Error {
+        let mut command = Command::new(shell.exe_path());
+        command.envs(exports);
+
+        let script = formatdoc! {"
+                # to avoid infinite recursion sourcing bashrc
```
Comment (ysndr) [inline cli/flox/src/commands/environment.rs:592]: do we need to apply the PATH fixup here as well?

File: cli/tests/environment-activate.bats [hunk 3]
```diff
@@ -167,11 +167,6 @@ env_is_activated() {
   # USER to REAL_USER.
   SHELL=bash USER="$REAL_USER" NO_COLOR=1 run -0 expect -d "$TESTS_DIR/activate/rc.exp" "$PROJECT_DIR"
   assert_output --partial "test_alias is aliased to \`echo testing'"
-
-  SHELL=bash USER="$REAL_USER" NO_COLOR=1 run "$FLOX_BIN" activate --dir "$PROJECT_DIR" -- type test_alias
-  assert_success
-  assert_output --partial "test_alias is aliased to \`echo testing'"
-
```
Comment (ysndr) [inline cli/tests/environment-activate.bats:174]: We don't run init scripts anymore so aliases set in the rc files are not available in the subshell

File: cli/flox/src/commands/environment.rs (hunk 2 above)
Comment (mkenigs) [reply cli/flox/src/commands/environment.rs:592]: Don't we bail if the environment is already active? So I don't think so?

Comment (ysndr) [reply cli/flox/src/commands/environment.rs:592]: i guess that's only relevant for something like

```
eval "$(flox activate -d a)"
tmux
flox activate -d b -- <uses something from a>
```

Comment (mkenigs) [reply cli/flox/src/commands/environment.rs:592]: Ah right. I guess that's the same case covered in `activate_inplace` (if a is already active and the user runs `flox activate -d a --in-place`, we'll call `reactivate_in_place`, not `activate_inplace`). But then we're not doing it in `activate_interactive`. I guess we should be consistent between `activate_inplace`, `activate_non_interactive`, and `activate_interactive`? None of those will be called if the environment is already activated. Any of them might be run when `a` has been re-ordered by path helper, and we're activating `b`, right?

File: cli/flox/src/commands/environment.rs [hunk 4]
```diff
@@ -768,7 +790,7 @@ impl Activate {
         Ok(())
     }
 
-    fn activate_non_interactive(
+    fn activate_inplace(
```
Comment (mkenigs) [inline cli/flox/src/commands/environment.rs:793]: nit: one method has `inplace` and the other has `in_place`

File: cli/flox/src/commands/environment.rs [hunk 5]
```diff
@@ -565,28 +565,57 @@ impl Activate {
...
+        let activate_error = if !self.run_args.is_empty() {
+            Self::activate_non_interactive(self.run_args, shell, exports, activation_path)
+        } else {
+            Self::activate_interactive(shell, exports, activation_path, now_active)
+        };
         // If we get here, exec failed!
         Err(activate_error)
     }
 
+    fn activate_non_interactive(
```
Comment (mkenigs) [inline cli/flox/src/commands/environment.rs:582]:
```suggestion
    /// Used for `flox activate -- run_args`
    fn activate_non_interactive(
```

File: cli/flox/src/commands/environment.rs (hunk 4 above)
Comment (mkenigs) [inline cli/flox/src/commands/environment.rs:793]:
```suggestion
    /// Used for `eval "$(flox activate)"`
    fn activate_inplace(
```

File: cli/flox/src/commands/environment.rs [hunk 6]
```diff
@@ -565,28 +565,57 @@ impl Activate {
...
+        activation_path: PathBuf,
+    ) -> anyhow::Error {
+        let mut command = Command::new(shell.exe_path());
+        command.envs(exports);
+
+        let script = formatdoc! {"
+                # to avoid infinite recursion sourcing bashrc
+                export FLOX_SOURCED_FROM_SHELL_RC=1
+
+                source {activation_path}/activate/{shell}
```
Comment (mkenigs) [inline cli/flox/src/commands/environment.rs:595]:
```suggestion
                // TODO: this script sets prompt, which isn't necessary
                source {activation_path}/activate/{shell}
```

**Review by mkenigs — CHANGES_REQUESTED** [review]
I think we need a consistent approach between `activate_inplace`, `activate_non_interactive`, and `activate_interactive` for path patching, but otherwise this looks good

**Comment by mkenigs** [comment]
> I think we need a consistent approach between `activate_inplace`, `activate_non_interactive`, and `activate_interactive` for path patching, but otherwise this looks good

@ysndr it doesn't look like you made those consistent, are you wanting to address that here or elsewhere?

**Comment by ysndr** [comment]
@mkenigs should be consistent now as per [307a1da](https://github.com/flox/flox/pull/889/commits/307a1daf13c24fe0343fd77c495adb506ec18375)

**Review by mkenigs — APPROVED** [review]
Would be good to have some tests for the path patching, but that wasn't the original goal of this PR so not going to block

**Pull request merged**